### DISCUSSION TITLE
Total CPU time increasing for high numbers of futures workers

### DISCUSSION THREAD

**Discussion by sciaba** [scikit-hep/coffea#876] [Q&A]
I noticed that if I run the AGC workload for increasing numbers of Python futures workers on a node with 128 CPU cores and reading from the local file system, the total CPU time for the entire process tree as measured by PrMon starts increasing for high numbers of workers (~30 or more). Ideally, it should remain constant, if the CPU time depended only on the amount of data to process. A similar thing happens with the RDF version, so it's not specific to Coffea, at least qualitatively (the causes might be different). Any idea about possible causes for this overhead?
I am attaching a plot comparing the CPU time to the processtime (as reported by Coffea) and to the wallclock time of the entire AGC workload multiplied by the number of workers:

![agc_coffea_cpu-proc_time](https://github.com/CoffeaTeam/coffea/assets/12764525/b58e950b-da24-446d-bfb9-69dcbca52a65)

**Comment by lgray** [comment]
@sciaba My first inclination is to think this has something to do with dask (if that's what you're using!) and the way to communicates with workers. It might be worth cross post on their github (with a bunch of details about the setup) and see if anyone comes back with some immediate answers.

**Reply by sciaba** [reply]
I'm not using Dask: I'm running the workload with USE_DASK = False, so I'm using the FuturesExecutor.

**Reply by lgray** [reply]
Ah, gotcha. I can't think of any right now that would cause total CPU time increases, at least from the mechanics of unpacking the file and manipulating the data. Is this all over local disk access or via xrootd?

**Reply by sciaba** [reply]
The plot above is for local disk access (RAID-0 of SSDs), but I just made a similar plot for access from EOS via xrootd.

![agc_coffea_eos_cpu-proc_time](https://github.com/CoffeaTeam/coffea/assets/12764525/053002ce-54d6-4977-98ad-7a2f203899b6)

while this plot compares the CPU time in the local disk access case with the one in the EOS xrootd case:

![agc_coffea_cpu_local_vs_eos](https://github.com/CoffeaTeam/coffea/assets/12764525/e9546022-498d-4ebd-a0e3-8372a2046a9f)

**Reply by lgray** [reply]
raid0 controller is software or hardware?

**Reply by sciaba** [reply]
It's software (with a 95% CL :) ).

**Reply by sciaba** [reply]
Just to add that by "CPU time" I mean the sum of user time and system time consumed by the process tree.

**Reply by lgray** [reply]
Might be that given the rather scattered nature of reading from root files. It might be interesting to try parquet/feather/rntuple and see if the scaling is still there. They're much more 'focused' when it comes to reading.

**Reply by lgray** [reply]
(it's really hard to fabricate a scenario where scheduling raid reads in the kernel starts taking up that much processing power though!)

**Reply by sciaba** [reply]
I was hypothesizing some bottleneck at the CPU level (maybe the cache is thrashing a lot?) but I'm not sure if this would generate additional CPU time (my knowledge of CPU architectures is simply insufficient....). Or maybe caching at the OS level...

**Reply by lgray** [reply]
Crank the CPU affinity way up and see what happens?

**Reply by lgray** [reply]
Data cache thrashing does make a bunch of sense.

**Reply by lgray** [reply]
Could also be TLB misses from context switching all the time. Pinning processes to cores would fix that.

**Reply by sciaba** [reply]
I'd be happy to try but I don't know how... (in case you can help me with that send me a DM not to pollute the thread!).

**Reply by lgray** [reply]
@sciaba `taskset -pc <cpu-number> <pid>` :-)
